### PR TITLE
feat(db_api): add an ability to set ReadOnly/ReadWrite connection mode

### PR DESCRIPTION
**Pull request by IlyaFaer**
Proposal for the user's question: https://github.com/cloudspannerecosystem/python-spanner-sqlalchemy/issues/97
Will require small changes in SQLAlchemy dialect as well.

### CODE REVIEW

File: google/cloud/spanner_dbapi/connection.py
Comment (IlyaFaer) [inline google/cloud/spanner_dbapi/connection.py:153]: It's done in the same way JDBC works: connection mode can be changed, but only if there is no active transaction on it.

File: google/cloud/spanner_v1/transaction.py
Comment (olavloite) [inline google/cloud/spanner_v1/transaction.py:109]: This will begin a read-only transaction on this connection. It is important to realize (and maybe also document) a couple of things when using a read-only transaction:
1. A read-only transaction will only read data at a specific point in time during the entire transaction. That means that when a user starts a read-only transaction, they will never see any new changes to the database until the read-only transaction has been closed and a new one begins. So if I start a new read-only transaction at time T1, and someone else executes and commits a read/write transaction at T2, the read-only transaction will still only see the data at T1 and not the data from T2. This can be surprising for many users, and it should be pointed out that most users who want to just read data should probably use autocommit mode, and not read-only transactional mode.
2. Read-only transactions cannot be committed or rolled back. You must therefore check the type of transaction that is in use when `commit` or `rollback` is called on the connection, and only do something with the transaction if it is a read/write transaction. (And of course return the session to the pool if it is a read-only transaction.)

File: tests/system/test_system_dbapi.py
Comment (olavloite) [inline tests/system/test_system_dbapi.py:450]: This should not raise an exception. This should remove the read-only transaction from the connection and return the session to the pool. If there is no possibility of 'committing' (or otherwise ending) a read-only transaction, the connection will __never__ see any new data after the read-only transaction has started.

File: tests/system/test_system_dbapi.py
Comment (olavloite) [inline tests/system/test_system_dbapi.py:430]: It would be good to add an extra system test that shows the following:
1. A read-only transaction can read data that was present in the database when the read-only transaction started.
2. If data is added using a separate read/write transaction (i.e. another user added data to the database), those records are __not__ visible to the read-only transaction in step 1.
3. __After__ committing/rolling back (closing) the read-only transaction, and starting a new read-only transaction, the data from step 2 should be visible to the read-only transaction.

**Review by olavloite — CHANGES_REQUESTED** [review]

File: tests/system/test_system_dbapi.py
Comment (IlyaFaer) [reply tests/system/test_system_dbapi.py:450]: It is something that comes from the backend with a message:
`400 Cannot commit or rollback a read-only or partitioned-dml transaction`. So, it's not us who raise this exception.

File: tests/system/test_system_dbapi.py
Comment (IlyaFaer) [reply tests/system/test_system_dbapi.py:430]: Done

File: google/cloud/spanner_v1/transaction.py
Comment (IlyaFaer) [reply google/cloud/spanner_v1/transaction.py:109]: Okay, I've added changes for item 2 and a system test for item 1.
Looks like we need to find a proper place for mentioning all of it. I don't see any docs for connection API...

File: google/cloud/spanner_dbapi/connection.py
Comment (olavloite) [inline google/cloud/spanner_dbapi/connection.py:55]:
```suggestion
        Flag to indicate that the connection may only execute queries and no update or DDL statements.
        If autocommit is true, the connection will use a single use read-only transaction with strong timestamp
        bound for each new statement, and will immediately see any changes that have been committed by
        any other transaction.
        If autocommit is false, the connection will automatically start a new multi use read-only transaction
        with strong timestamp bound when the first statement is executed. This read-only transaction will be
        used for all subsequent statements until either commit() or rollback() is called on the connection. The
        read-only transaction will read from a consistent snapshot of the database at the time that the
        transaction started. This means that the transaction will not see any changes that have been
        committed by other transactions since the start of the read-only transaction. Commit or rolling back
        the read-only transaction is semantically the same, and only indicates that the read-only transaction
        should end a that a new one should be started when the next statement is executed.
```

File: google/cloud/spanner_dbapi/connection.py
Comment (olavloite) [inline google/cloud/spanner_dbapi/connection.py:137]:
```suggestion
                True if the connection may only be used for database reads.
```

File: google/cloud/spanner_v1/transaction.py
Comment (olavloite) [inline google/cloud/spanner_v1/transaction.py:89]:
```suggestion
            (Optional) If True, a read-only transaction with strong timestamp bound will be started.
            Otherwise a read/write transaction will be started.
```

File: google/cloud/spanner_v1/transaction.py
Comment (olavloite) [reply google/cloud/spanner_v1/transaction.py:109]: Thanks, this looks great. I've added a couple of suggestion for documentation in the code.

Also, there is one more thing that we could simplify for read-only transactions: A read-only transaction will never be aborted by Cloud Spanner, so keeping track of a running checksum for the returned result sets is not necessary.  We could therefore skip that when we see that the transaction is in read-only mode.

File: google/cloud/spanner_dbapi/connection.py
Comment (olavloite) [inline google/cloud/spanner_dbapi/connection.py:220]: I think it would be better to change the `except Aborted` blocks in the different methods so that they never call `retry_transaction` when it is a read-only transaction. Instead the `Aborted` error should just be propagated to the caller when it is a read-only transaction (and in theory it should also never happen).

File: google/cloud/spanner_dbapi/cursor.py
Comment (olavloite) [inline google/cloud/spanner_dbapi/cursor.py:296]: Add an `if` statement to this and other `except Aborted` blocks to prevent `retry_transaction` from being called when it is a read-only transaction.

**Review by larkee — CHANGES_REQUESTED** [review]
Read-only transaction in the Python client are represented by the `Snapshot` [class](https://github.com/googleapis/python-spanner/blob/master/google/cloud/spanner_v1/snapshot.py#L532).

Please do not modify `transaction.py`. Use `Snapshot` instead.

**Comment by IlyaFaer** [comment]
@larkee, hm, okay. If we want to use the `Snapshot()` class, we probably could re-use the existing method for single-use reads. Let me know if you think it's better to use a snapshot as a transaction for several sequential reads.

**Comment by larkee** [comment]
> @larkee, hm, okay. If we want to use the `Snapshot()` class, we probably could re-use the existing method for single-use reads. Let me know if you think it's better to use a snapshot as a transaction for several sequential reads.

By default, a snapshot is a single-use read-only transaction so it would only be valid for one read operation. However, there is an option make a multi-use snapshot in which all reads are from the same snapshot of data i.e. data is consistent.

Multi-use snapshot example: https://cloud.google.com/spanner/docs/transactions#ro_transaction_example
Single-use snapshot example: https://cloud.google.com/spanner/docs/reads#single_read_methods

With this in mind, I think using `multi_use` snapshots should only be used when `read_only=True` and `AUTOCOMMIT=False`. WDYT?

**Comment by olavloite** [comment]
> With this in mind, I think using `multi_use` snapshots should only be used when `read_only=True` and `AUTOCOMMIT=False`. WDYT?

Yes, that is the intention here. So the implementation should roughly be:

1. `read_only=True` and `auto_commit=False`: Use `multi_use` snapshots and only allow queries/reads. Calling `commit()` or `rollback()` should close the current multi-use snapshot, and the next statement that is executed should create a new multi-use snapshot.
2. `auto_commit=True` (and regardless of the state of `read_only`): Always use single-use snapshots for queries/reads.

**Review by larkee — COMMENTED** [review]
LGTM, however I would like some system tests to confirm it is WAI

File: google/cloud/spanner_dbapi/connection.py
Comment (olavloite) [inline google/cloud/spanner_dbapi/connection.py:232]: This should not be necessary. The _rerun_previous_statements method should never be called when we are in read-only mode (and also not when in auto-commit mode).

Re-running a list of statements is only necessary when we are in a multi-use read/write transaction.

File: google/cloud/spanner_dbapi/connection.py
Comment (olavloite) [inline google/cloud/spanner_dbapi/connection.py:241]: Here also: This should not be necessary. We should never call _rerun_previous_statements unless the transaction is a multi-use read/write transaction.

File: google/cloud/spanner_dbapi/connection.py
Comment (olavloite) [reply google/cloud/spanner_dbapi/connection.py:290]: I think it would be good in that case to add a check at the beginning of the method that actually checks that the state of the connection is `read_only==True && autocommit==False`. That prevents any accidental errors in the future when making changes.

File: google/cloud/spanner_dbapi/cursor.py
Comment (olavloite) [inline google/cloud/spanner_dbapi/cursor.py:408]: Doesn't this mean that the statement was an update statement? In that case, it should not be possible to ever enter this if-block, as an update statement cannot be executed on a read-only snapshot.
Or is this some specific handling for `SELECT COUNT(*)` statements?

File: google/cloud/spanner_dbapi/cursor.py
Comment (olavloite) [inline google/cloud/spanner_dbapi/cursor.py:427]: A read-only snapshot will never be aborted by Cloud Spanner, so this is not necessary.

File: google/cloud/spanner_dbapi/cursor.py
Comment (IlyaFaer) [reply google/cloud/spanner_dbapi/cursor.py:408]: I don't think it's a relevant part of code at all, it was probably copy-pasted from somewhere else. `Snapshot.execute_sql()` returns `StreamedResultSet`, and at the first look I don't see how an `int` can get here.

Still, in this PR I wasn't going to change anything, I've only rewritten the method to make it able to pass a snapshot as an arg (not only use it as a context). So, there is a note for this `if` statement, but let's keep it for another PR.

File: google/cloud/spanner_dbapi/cursor.py
Comment (IlyaFaer) [reply google/cloud/spanner_dbapi/cursor.py:427]: Yeah, it also looks like something copied from another part of code (where the `execute_sql()` method is called not with `Snapshot`). Will take a look.

File: google/cloud/spanner_dbapi/cursor.py
Comment (olavloite) [reply google/cloud/spanner_dbapi/cursor.py:352]: This comment has been marked as resolved, but no changes were made. Do you disagree with my suggestion, or is this something you are still working on?

File: google/cloud/spanner_dbapi/connection.py
Comment (larkee) [inline google/cloud/spanner_dbapi/connection.py:318]: I think it would make more sense to add the `read_only` check here and exit early. If the connection is read only, then none of the code below should be relevant. WDYT?

This also applies to `rollback()`.

File: google/cloud/spanner_dbapi/connection.py
Comment (larkee) [reply google/cloud/spanner_dbapi/connection.py:287]: This was marked as resolved but no changes were made. Please address.

File: google/cloud/spanner_dbapi/connection.py
Comment (olavloite) [reply google/cloud/spanner_dbapi/connection.py:318]: The `self._release_session()`call would also be needed in the early exit branch in that case.

File: google/cloud/spanner_dbapi/connection.py
Comment (IlyaFaer) [reply google/cloud/spanner_dbapi/connection.py]: Done

**Comment by skuruppu** [comment]
@IlyaFaer is this ready for another round of reviews?

**Comment by IlyaFaer** [comment]
@skuruppu, yeah, I think everything is fixed. The only unresolved question was answered by Knut, so I assume it's also resolved.

**Comment by olavloite** [comment]
@IlyaFaer 

There are a couple of comments that were marked as resolved, but with no action taken, such as:
* https://github.com/googleapis/python-spanner/pull/475/files#r689372738
* https://github.com/googleapis/python-spanner/pull/475/files#r684914316

Would you mind taking another look at those?

Also, @larkee asked for a system test to be added to show the working of the feature. Could you take a look at that as well?

**Comment by IlyaFaer** [comment]
> There are a couple of comments that were marked as resolved, but with no action taken, such as:
https://github.com/googleapis/python-spanner/pull/475/files#r689372738
https://github.com/googleapis/python-spanner/pull/475/files#r684914316

I've written a comment for all the related comments: https://github.com/googleapis/python-spanner/pull/475#discussion_r689369342
So, I'm going to work on it, but in a separate PR, as it doesn't seem to be related to this one, but will require not the smallest changes.

The system test has been lost! It was actually added, but because of (not mine) big system tests refactoring PR it was erased during merge.
https://github.com/googleapis/python-spanner/pull/475/commits/b8eefed65d660dd7cd3cef27d5a61f5caf18edad

**Comment by olavloite** [comment]
> I've written a comment for all the related comments: [#475 (comment)](https://github.com/googleapis/python-spanner/pull/475#discussion_r689369342)

That is true for https://github.com/googleapis/python-spanner/pull/475/files#r689320974, but not for the two other comments in this case which are directly related to this change:
1. This PR introduces support for multi-use read-only transactions.
2. The existing read/write transactions use a checksum and retry mechanism to handle aborted errors.
3. The newly introduced read-only transactions now also use the checksum and retry mechanism that was in place for read/write transactions.
4. Read-only transactions however will never be aborted by Cloud Spanner, which means that keeping track of a checksum and checking for aborted errors for read-only transactions is not necessary.

So I think this PR should take into account that read-only transactions will never be aborted by Cloud Spanner, they should not keep track of a result set checksum, and they should not be retried in case of an aborted error.

**Comment by IlyaFaer** [comment]
@olavloite, the PR is adding ReadOnly transactions support, there are users, who wait for this feature. The error with `Aborted` expectation was always there - it's not optimal, yes, but lived with it and we can live little bit more, while lack of ReadOnly transactions support can block users' projects. As I see it, we should merge this PR ASAP (it's been 1.5 months in progress already), so that people at least could use the package. The error will be fixed anyway, it just should not delay more significant changes (and hence customers' projects).

**Comment by skuruppu** [comment]
I would be a bit worried about doing the extra performance cost of keeping track of checksums for read-write transactions though. I wouldn't be comfortable releasing that. Is it hard to fix @IlyaFaer?

**Comment by IlyaFaer** [comment]
@skuruppu, of course, it's not. I don't want to sound boring, but I can't see why you both think it's that performance influencing. We don't calculate checksums in `read_only` - the `if`s are already added (https://github.com/googleapis/python-spanner/pull/475/commits/8332d4977e658d7a47971932c1d19502bcb9d7fb). As for retrying `read_only`, well, `read_only` transactions don't fail with `Aborted`, so they're not gonna be retried anyway. Thus, I don't see why these changes should be urgent and delay other fixes...

Anyway it's pushed.

**Comment by olavloite** [comment]
> ... but I can't see why you both think it's that performance influencing. We don't calculate checksums in read_only - the ifs are already added (8332d49).

Sorry, that was my fault. I didn't realize that you had already created that change, as the [comment that you referenced](https://github.com/googleapis/python-spanner/pull/475#discussion_r689369342) above didn't say anything about that.

> ... Thus, I don't see why these changes should be urgent and delay other fixes...

In my opinion, the current implementation would have been slightly confusing for a potential new developer. The retry mechanism is not needed for read-only transactions. I think that it is better that it is completely clear in the code to prevent anyone from accidentally re-enabling it in the future. (This last commit for example has also removed an unnecessary test case that tested the retry mechanism for a query in autocommit mode; a situation that could not happen, but still was part of the code base.)

Anyways, I'm happy enough with this as-is, but we should add a system test ASAP to actually show that it is working as intended. If @skuruppu and/or @larkee are OK with that, then we could do that in a separate PR.

**Comment by skuruppu** [comment]
Ultimately @larkee needs to approve this as the code owner so please do add any tests that they ask for before merging.

**Comment by larkee** [comment]
@IlyaFaer I second what @olavloite said. The refactor is less about performance (since it should be roughly equivalent) but is more about readability and reducing confusion for new developers.

However, if you create a tracking issue for the refactor work, I will approve and merge this PR :+1:

**Comment by IlyaFaer** [comment]
I think all comments are already processed.

I've pushed changes with adding `if` statements into `fetch` methods to avoid retrying `read_only` transactions, etc. (see https://github.com/q-logic/python-spanner/commit/ac8c4b2b62cd228d687decda4ec54b32553164e4). So, I don't think any refactoring left here.

I also returned back that deleted-by-merge-of-another-commit system test. It checks that `UPDATE` operation will cause an error, while `SELECT` will work fine.

I think it's worth adding a unit test, which'll check that `read_only` transactions are not retried on `Aborted` exception. Otherwise, I think we don't have anything for a separate issue.

**Review by larkee — APPROVED** [review]

**Review by olavloite — APPROVED** [review]

**Comment by IlyaFaer** [comment]
Looking at `automerge` label set by googler, I suppose it's not a problem if I'll merge it by myself (while tests are green and there are no changes in the `main` branch).

**Pull request merged**
